src/ivc/protogalaxy/mod.rs

```diff
@@ -507,6 +514,65 @@ mod verify_chip {
             .map_err(|err| Error::BetasStroke { err })
     }
 
+    /// Evaluate the values of the Lagrange polynomial zero for a cyclic subgroup of length `n` (`2.pow(log_n)`) at
+    /// the `challenge` point
+    ///
+    /// You can look at [`fft::get_omega_or_inv`] to see how the target cyclic group is formed
+    ///
+    /// # Mathematical Representation
+    ///
+    /// ```math
+    /// L_i(X)=\frac{\omega^i}{n}\frac{X^n-1}{X-\omega^i}
+    /// ```
+    /// where {1, \omega, \omega^2, ..., \omega^n} - cyclic group, check [`iter_cyclic_subgroup`] for
+    /// more details
+    ///
+    /// # Generics
+    /// `T` is setup for main gate
+    /// - `L`: 'Length' - constant representing the number of instances to
+    ///                   fold in a single `prove`. `L-1` be power of two
+    fn eval_lagrange_zero_poly<F: PrimeField, const T: usize, const L: usize>(
+        region: &mut RegionCtx<F>,
+        main_gate: &MainGate<F, T>,
+        cha: &mut ValuePowers<F>,
+    ) -> Result<AssignedValue<F>, Halo2PlonkError> {
+        let lagrange_domain = PolyContext::<F>::get_lagrange_domain::<L>();
+        let points_count = 2usize.pow(lagrange_domain);
+
+        let inverted_n = F::from_u128(points_count as u128)
+            .invert()
+            .expect("safe because it's `2^log_n`");
+        let value = iter_cyclic_subgroup::<F>(lagrange_domain).next().unwrap();
+
+        let X = cha.value();
+
+        let X_sub_value = main_gate.add_with_const(region, &X, -value)?;
+
+        let (is_zero_X_sub_value, X_sub_value_inverted) =
+            main_gate.invert_with_flag(region, X_sub_value)?;
+
+        let X_pow_n = cha.get_or_eval(region, main_gate, points_count)?;
+        let X_pow_n_sub_1 = main_gate.add_with_const(region, &X_pow_n, -F::ONE)?;
+
+        let is_zero_X_pow_n_sub_1 = main_gate.is_zero_term(region, X_pow_n_sub_1.clone())?;
+
+        let is_numerator_denominator_zero =
+            main_gate.mul(region, &is_zero_X_sub_value, &is_zero_X_pow_n_sub_1)?;
+
+        let lhs = main_gate.mul(region, &X_pow_n_sub_1, &X_sub_value_inverted)?;
+        let fractional = main_gate.mul_by_const(region, &lhs, value * inverted_n)?;
+
+        let zero = region.assign_advice(
+            || "zero",
+            main_gate.config().state[0],
+            Halo2Value::known(F::ZERO),
+        )?;
+
+        region.next();
+
+        main_gate.conditional_select(region, &zero, &fractional, &is_numerator_denominator_zero)
```

Comment:
The logic here is not the same as:

```
 let X_sub_value_inverted = X.sub(value).invert();
            let X_pow_n_sub_1 = X.pow([points_count as u64]) - F::ONE;

            // During the calculation, this part of the expression should be reduced to 1, but we
            // get 0/0 here, so we insert an explicit `if`.
            if X_pow_n_sub_1.is_zero_vartime() && X_sub_value_inverted.is_none().into() {
                F::ONE
            } else {
                value * inverted_n * (X_pow_n_sub_1 * X_sub_value_inverted.unwrap())
            }
        })
```

Reply:
You right, should be one, not zero
Will fix